### ISSUE TITLE
Mono erroneously falls back to old tools

### ISSUE THREAD

**Issue by icedream** [icedream/citizenmp-server-updater#1]
Updater version: 1.0.3

OS information:
Linux Arch.

Mono version:
```
Mono JIT compiler version 3.10.0 (tarball Mon Oct  6 21:13:11 UTC 2014)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  x86
        Disabled:      none
        Misc:          softdebug
        LLVM:          supported, not enabled.
        GC:            sgen
```
Console log (http://pastebin.com/tGay1Txz):
```

[ztjuh@ztjuh citimp_upd]$ mono citimp_upd.exe Server
Updating source code...
Downloading dependencies...
Patching assembly information...
Updating SSL certificates for NuGet...
Updating NuGet...
Building server binaries...
/home/ztjuh/citimp_upd/src/Vendor/npsharp/src/client/NPSharp.Client.csproj:  warning : Project has unknown ToolsVersion '12.0'. Using the default tools version '2.0' instead.
: error : Error initializing task MsBuild: Cannot cast from source type to destination type.
: error : Error building target CheckPrerequisites: Object reference not set to an instance of an object
/home/ztjuh/citimp_upd/src/Vendor/npsharp/src/client/NPSharp.Client.csproj: error : Object reference not set to an instance of an object
/home/ztjuh/citimp_upd/src/CitizenMP.Server.sln: error : Error executing task MSBuild: Object reference not set to an instance of an object
: error : Error building target Build: Object reference not set to an instance of an object

Unhandled Exception:
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.NullReferenceException: Object reference not set to an instance of an object
  at Microsoft.Build.BuildEngine.ConsoleLogger+BuildRecord.TaskFinishedHandler (Microsoft.Build.Framework.TaskFinishedEventArgs args) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.ConsoleLogger.TaskFinishedHandler (System.Object sender, Microsoft.Build.Framework.TaskFinishedEventArgs args) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.BuildEvent.ExecuteFinishedHandler (Microsoft.Build.Framework.BuildStatusEventArgs finished_args) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.ConsoleLogger+BuildRecord.PopEvent[ProjectFinishedEventArgs] (Microsoft.Build.Framework.ProjectFinishedEventArgs finished_args) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.ConsoleLogger.PopEvent[ProjectFinishedEventArgs] (System.Object sender, Microsoft.Build.Framework.ProjectFinishedEventArgs args) [0x00000] in <filename unknown>:0
  at (wrapper delegate-invoke) <Module>:invoke_void_object_ProjectFinishedEventArgs (object,Microsoft.Build.Framework.ProjectFinishedEventArgs)
  at Microsoft.Build.BuildEngine.EventSource.FireProjectFinished (System.Object sender, Microsoft.Build.Framework.ProjectFinishedEventArgs pfea) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.Engine.LogProjectFinished (Microsoft.Build.BuildEngine.Project project, Boolean succeeded) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.Engine.EndProjectBuild (Microsoft.Build.BuildEngine.Project project, Boolean succeeded) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.Project.Build (System.String[] targetNames, IDictionary targetOutputs, BuildSettings buildFlags) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.Project.Build (System.String[] targetNames, IDictionary targetOutputs) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.Project.Build (System.String[] targetNames) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.Project.Build () [0x00000] in <filename unknown>:0
  at CitizenMP.Server.Installer.Program.Build (System.String projectFilePath, IDictionary`2 buildProperties, LoggerVerbosity verbosity, System.String logPath) [0x00000] in <filename unknown>:0
  at CitizenMP.Server.Installer.Program.Main (System.String[] args) [0x00000] in <filename unknown>:0
  at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0
  --- End of inner exception stack trace ---
  at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0
  at System.Reflection.MethodBase.Invoke (System.Object obj, System.Object[] parameters) [0x00000] in <filename unknown>:0
  at Program.Main (System.String[] args) [0x00000] in <filename unknown>:0
[ERROR] FATAL UNHANDLED EXCEPTION: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.NullReferenceException: Object reference not set to an instance of an object
  at Microsoft.Build.BuildEngine.ConsoleLogger+BuildRecord.TaskFinishedHandler (Microsoft.Build.Framework.TaskFinishedEventArgs args) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.ConsoleLogger.TaskFinishedHandler (System.Object sender, Microsoft.Build.Framework.TaskFinishedEventArgs args) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.BuildEvent.ExecuteFinishedHandler (Microsoft.Build.Framework.BuildStatusEventArgs finished_args) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.ConsoleLogger+BuildRecord.PopEvent[ProjectFinishedEventArgs] (Microsoft.Build.Framework.ProjectFinishedEventArgs finished_args) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.ConsoleLogger.PopEvent[ProjectFinishedEventArgs] (System.Object sender, Microsoft.Build.Framework.ProjectFinishedEventArgs args) [0x00000] in <filename unknown>:0
  at (wrapper delegate-invoke) <Module>:invoke_void_object_ProjectFinishedEventArgs (object,Microsoft.Build.Framework.ProjectFinishedEventArgs)
  at Microsoft.Build.BuildEngine.EventSource.FireProjectFinished (System.Object sender, Microsoft.Build.Framework.ProjectFinishedEventArgs pfea) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.Engine.LogProjectFinished (Microsoft.Build.BuildEngine.Project project, Boolean succeeded) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.Engine.EndProjectBuild (Microsoft.Build.BuildEngine.Project project, Boolean succeeded) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.Project.Build (System.String[] targetNames, IDictionary targetOutputs, BuildSettings buildFlags) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.Project.Build (System.String[] targetNames, IDictionary targetOutputs) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.Project.Build (System.String[] targetNames) [0x00000] in <filename unknown>:0
  at Microsoft.Build.BuildEngine.Project.Build () [0x00000] in <filename unknown>:0
  at CitizenMP.Server.Installer.Program.Build (System.String projectFilePath, IDictionary`2 buildProperties, LoggerVerbosity verbosity, System.String logPath) [0x00000] in <filename unknown>:0
  at CitizenMP.Server.Installer.Program.Main (System.String[] args) [0x00000] in <filename unknown>:0
  at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0
  --- End of inner exception stack trace ---
  at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0
  at System.Reflection.MethodBase.Invoke (System.Object obj, System.Object[] parameters) [0x00000] in <filename unknown>:0
  at Program.Main (System.String[] args) [0x00000] in <filename unknown>:0
```

Originally reported by Ztjuh.

**Comment by icedream** [comment]
Bug confirmed with [latest develop build](https://ci.icedream.kthx.at/job/citiserver-updater-develop/lastBuild/) on Debian Wheezy. So I guess upgrading to .NET Framework 4.5 didn't really work out...

**Comment by icedream** [comment]
Found out how xbuild solves their issue. They have a [generated assembly configuration](https://raw.githubusercontent.com/mono/mono/master/mcs/tools/xbuild/data/xbuild.exe.config.in) for each target that makes all assemblies redirect to the compatible version.

This should be solvable in our case without having to use a .config file, instead we can set up a custom AssemblyResolve handler.

**Comment by icedream** [comment]
Build [unstable.58](https://ci.icedream.kthx.at/job/citiserver-updater-develop/55/) contains the fix and seems to be working fine, at least on my virtual machine. I'll send Ztjuh a message so he can test it before this gets merged into 1.1

**Comment by icedream** [comment]
Issue fixed in version 1.1.
